### ISSUE TITLE
Stop an animation

### ISSUE THREAD

**Issue by Alf21** [juliangarnier/anime#188]
Is it possible to stop an animation from inside of another animation?

I created two timelines. The first one is infinite and the second one has two animations.
After finishing animation 2, i want to stop animation 1 and go to the next step.

Thats currently not possible i think, so i tried another thing:
I set the autoplay of anim1 to false and started it with 'anim1.play()' when anim2.begin callback is called.
Then im waiting for callback anim2.complete to create a third animation.

But it is also buggy, if i set an animation to autoplay = false, i can't start it from any event or anything else...

Any solutions?

**Comment by juliangarnier** [comment]
Please provide some code.

**Comment by pakx** [comment]
I have a similar "issue" -- is there a way to "cancel" a currently-playing animation? I didn't notice anything in the docs, or on looking at the properties of the return value of `anim(...)`.

(Separately, awesome library -- many thanx!)

**Comment by snc** [comment]
@pakx You can call `pause()` on the return value of `anim(...)`.

**Comment by pakx** [comment]
@snc, yes, I saw `pause()` ... but I assume that only "pauses" an animation, possibly to be continued later, correct? What if I want to cancel/kill it completely, including having it release whatever resources it's holding (timers, memory allocation, etc)?

**Comment by wintercounter** [comment]
I'm pretty much amazed that there is no option for this.  I'm converting my animation from TweenMax. It had `TweenMax.killTweensOf(el);`. I have multiple hover animations and i want to stop all the non-hovered elements from animating.

**Comment by juliangarnier** [comment]
@pakx @wintercounter See anime.remove(el) https://github.com/juliangarnier/anime/#animeremovetarget

**Comment by ericmorand** [comment]
@juliangarnier what if I only want to stop one of the animations and not all of them? pause() will only pause the animation and let it hanging around in memory.

**Comment by juliangarnier** [comment]
I don't think it's a problem.
Have you noticed performance issues using .pause()?

**Comment by akash-pal** [comment]
pause() doesn't work for me after i start the animation

**Comment by garrettmaring** [comment]
If you trigger `play` after `pausing` will it resume from the point where it was paused? This indicates to me that there is a distinct use for a `cancel`.

**Comment by Cristy94** [comment]
Any news about this? There is no `animation.stop()` or `animation.destroy()` ?

**Comment by zortext** [comment]
> @pakx @wintercounter See anime.remove(el) https://github.com/juliangarnier/anime/#animeremovetarget

What if we are not animating a DOM element?
```
//Make an Object counter
var data = {count: 0};
anime({
   targets: data,
   count: 1000,                          
   update:myUpdateFunc
 });
```
anime.remove method does not work. I am also looking for a functionality like Mootools Fx.cancel 

https://mootools.net/core/docs/1.6.0/Fx/Fx#Fx:cancel

**Comment by AaronBeaudoin** [comment]
It seems that removing animations from the `anime.running` array can effectively cancel them.

Setting `anime.running.length` to `0` seems to stop all animations. It seems you can also pop animations from the array and they will stop.

It appears that for popped animations `animation.completed` will remain `false` and `animation.finished` will remain unresolved, likely because `animation.currentTime`, and `animation.progress` become frozen. The animation seems effectively dead, AKA cancelled or stopped.

I don't know if this is the intended way to do this, but I'm pretty sure its effects are what we are all looking for here. Hopefully there are no unseen side effects.

@juliangarnier If this is how we are supposed to do this could you add it to the docs? :)

**Comment by jameskerr** [comment]
If you want to cancel an animation with only the animation instance, I wrote a function that will gather all the targets, to then call `anime.remove` on each one. However, it would be much easier with an api as suggested above like `instance.cancel()`.

```js
function getTargets(ani) {
  return ani.children.reduce(
    (all, one) => all.concat(getTargets(one)),
    ani.animatables.map((a) => a.target)
  )
}

function cancel(ani) {
  getTargets(ani).forEach(anime.remove)
}
 ```
```js
let ani = anime.timeline(...).add(...).add(...)

// Later on...

cancel(ani)
```

**Comment by lukejagodzinski** [comment]
I'm surprised that this is not implemented yet. 2 years old issue and solution is pretty simple.
I have dozens of animations that stay paused forever in the queue without being able to cancel/stop/destroy them which takes memory. The more I use my app the more resources it takes. It can't garbage collect them when instances are being used in the array of paused instances.

@juliangarnier please implement this feature

**Comment by SupahSprinkle** [comment]
I stumbled upon this looking for something else. 
I just started using anime.js for a class and needed to stop an animation. Eventually this animation will be triggered off of a check box.
![RemoveAnim](https://user-images.githubusercontent.com/31256571/69775486-dcfb1400-1166-11ea-9286-28f23fbb35ab.gif)

I remove the animation by essentially removing the object that is being animated. Sort of hacky, but it works like a charm.
![image](https://user-images.githubusercontent.com/31256571/69775389-96a5b500-1166-11ea-9bad-d62eeb869255.png)

**Comment by lukejagodzinski** [comment]
Actually solution with `anime.remove(el)` kinda works but from API perspective it's ugly. It should be done on the instance level as this one removes element from all the instances which for some people might be wrong

**Comment by polymer940c** [comment]
You can use `.pause()` 
I had to update to new version also
here's the CDN: `https://cdn.jsdelivr.net/npm/animejs@3.0.1/lib/anime.min.js`
https://animejs.com/documentation/#playPause

**Comment by dangelion** [comment]
Any news on a `destroy` like method? I just can't remove the target element.

**Comment by mracette** [comment]
The issue with `anime.remove(el)` is that an attached callback such as `change` will continue running even with no targets. So it's not a true stop.

**Comment by wintercounter** [comment]
This still keeps popping in my life, now with a React project. It's causing huge memory leaks. Simply having `setState` in any animejs callback will force the browser to keep my component in memory and there is just no way to clean up created timelines...

**Comment by yunusbayraktaroglu** [comment]
You need to store animation to pause.

var pausableAnimation = anime({ ... });
pausableAnimation.start();
pausableAnimation.pause();

**Comment by akbr** [comment]
Formalizing and bumping @AaronBeaudoin 's https://github.com/juliangarnier/anime/issues/188#issuecomment-522206934, which seems to be the best we have right now:

```js
function cancelAnimation (animation) {
  let activeInstances = anime.running;
  let index = activeInstances.indexOf(animation);
  activeInstances.splice(index, 1);
}
```

Note the caveats about `complete` (will stay `false`) and `finished` not resolving.

**Comment by yellow1912** [comment]
I really hope we have a destroy method on the anime instance. Is there any possibility?

**Comment by xJkit** [comment]
`anime.destory()` is necessary.

**Comment by Cristy94** [comment]
Why is this closed if the issue is not solved?

**Comment by baba43** [comment]
This is pretty essential for properly integrating this into frameworks like React, isn't it?

**Comment by Gameghostify** [comment]
+1, this would be great to have

**Comment by yougotashovel** [comment]
Finding when a React component remounts, the animation timeline gets glitchy.  Having a destroy option like GSAP would help a lot.

Then I could remove the animation on unmount.

**Comment by juliangarnier** [comment]
Will work on that for the next release!

**Comment by xiziliang** [comment]
Looking forward to`anime.destory`

**Comment by juliangarnier** [comment]
Hey everyone!
Calling [`animation.pause()`](https://animejs.com/documentation/animation/animation-methods/pause) now properly removes the animation from memory. This is the preferred way to visually "stop" an animation from running.
If you want to completely destroy an animation, you can now call [`animation.revert()`](https://animejs.com/documentation/animation/animation-methods/revert), this stops the animation and reverts the styles / properties changes applied by the animation to the targets.
To remove an specific animation / target or tween from a timeline, you can now call [`timeline.remove()`](https://animejs.com/documentation/timeline/timeline-methods/remove).
To remove one or multiple targets from all active animations, a specific instance or a specific property, you can now use the updated [`utils.remove()`](https://animejs.com/documentation/utilities/remove) function.

Closing for now since I think this issue tries to address many different things.
Feel free to open a new one if you think the APIs listed above don't cover a specific use case.
Thanks
